### PR TITLE
🧠 Logic: 🗂️ source_file/1 predicate

### PR DESCRIPTION
**Pull request by amimart**
Implements #308 

Refer to the issue concerning the predicate definition.

Regarding the implementation, to manage only files loaded as prolog sources a trick has been done using reflection on the prolog VM.

## 🎯 Example

```sh
> okp4d --home $CHAIN_HOME query logic ask --program ":- consult('cosmwasm:okp4-storage:okp41ffzp0xmjhwkltuxcvccl0z9tyfuu7txp5ke0tpkcjpzuq9fcj3pqrteqt3?query=%7B%22object_data%22%3A%7B%22id%22%3A%221cc6de7672c97db145a3940df2264140ea893c6688fa5ca55b73cb8b68e0574d%22%7D%7D')." "source_file(X)."
answer:
  has_more: false
  results:
  - substitutions:
    - term:
        arguments: []
        name: '''cosmwasm:okp4-storage:okp41ffzp0xmjhwkltuxcvccl0z9tyfuu7txp5ke0tpkcjpzuq9fcj3pqrteqt3?query=%7B%22object_data%22%3A%7B%22id%22%3A%221cc6de7672c97db145a3940df2264140ea893c6688fa5ca55b73cb8b68e0574d%22%7D%7D'''
      variable: X
  success: true
  variables:
  - X
gas_used: "70002"
height: "2397"
```

### CODE REVIEW

File: x/logic/predicate/file.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		result = append(result, engine.NewAtom(filename))
+	}
+
+	return engine.Unify(
+		vm,
+		file,
+		engine.List(result...),
+		cont,
+		env,
+	)
```
[the remark this answers is not in the document]
Comment (amimart) [reply x/logic/predicate/file.go:37]: You're totally right!

**Review by bdeneux — APPROVED** [review]
Very nice 😁 That seems good to me 👍

File: x/logic/predicate/file_test.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+			},
+			{
+				query:       "source_file(X).",
+				wantSuccess: false,
+			},
+			{
+				query:       "consult(file1), consult(file2), source_file(X).",
+				wantResult:  []types.TermResults{{"X": "file1"}, {"X": "file2"}},
+				wantSuccess: true,
+			},
```
Comment (ccamel) [inline x/logic/predicate/file_test.go:55]: If the returned sources are sorted, add a test case which shows it. For example:
```go
          {
				query:       "consult(file2), consult(file3), consult(file1), source_file(X).",
				wantResult:  []types.TermResults{{"X": "file1"}, {"X": "file2"}, {"X": "file3"}},
				wantSuccess: true,
			},
```

File: x/logic/predicate/file.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	return engine.Delay(promises...)
+}
+
+func getLoadedSources(vm *engine.VM) map[string]interface{} {
+	loadedField := reflect.ValueOf(vm).Elem().FieldByName("loaded").MapKeys()
+	loaded := make(map[string]interface{}, len(loadedField))
+	for _, value := range loadedField {
+		loaded[value.String()] = nil
```
Comment (ccamel) [inline x/logic/predicate/file.go:64]: I usually use a `map[string]struct{}` to represent sets, with `struct{}{}` as an `empty` value (unique and named). But this is a matter of taste.

**Review by ccamel — CHANGES_REQUESTED** [review]

**Review by ccamel — APPROVED** [review]
All good. Great improvement.

**Pull request merged**